sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala

```diff
@@ -274,7 +274,10 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
           className = className,
           isTemporary = true)
 
-      case _ => throw QueryCompilationErrors.noSuchFunctionError(ident, plan)
+      case _ =>
+        val catalogPath = (currentCatalog +:
+          sparkSession.sessionState.catalogManager.currentNamespace).mkString(".")
+        throw QueryCompilationErrors.unresolvedRoutineError(ident, Seq(catalogPath), plan.origin)
```

Comment:
This should be no different than https://github.com/apache/spark/pull/39095/files#diff-583171e935b2dc349378063a5841c5b98b30a2d57ac3743a9eccfe7bffcb8f2aR173

To avoid duplicating code, we can even do
```
case other =>
  analyzer.checkAnalysis(other)
```

Reply:
> analyzer.checkAnalysis(other)

We must return a `Function` at the end. Do you propose to construct a fake `Function` or throw some exception (which one?)?